### DISCUSSION TITLE
Requiring the `PStore` adapter (and a few others) changes the default adapter

### DISCUSSION THREAD

**Discussion by offbyone** [jnunemaker/flipper#690] [General]
When I require a `pstore` adapter in a library in my rails app that isn't loaded automatically at engine initialization, the three lines at the bottom are replacing the global instance adapter with the newly loaded adapter. That's sort of counter-intuitive.

Is this intentional behaviour? If it is, then how do I require `flipper/adapters/pstore` _without_ changing my global adapter?

**Comment by bkeepers** [comment]
Hey @offbyone, this is indeed intentional. Each adapter will configure itself as the default when it is loaded. If your app is loading more than one adapter, I would recommend being explicit about which one you want to use:

```ruby
# config/initializers/flipper.rb
require "flipper/adapters/pstore"
require "flipper-active_record"

Flipper.configure do |config|
  config.adapter { Flipper::Adapters::ActiveRecord.new }
end
```

> That's sort of counter-intuitive.

We thought it was worth the tradeoff to make it "just work". I'd love to hear a counter-argument.

**Comment by offbyone** [comment]
One reason it’s not so great is that the act of importing the pstore adapter creates a config file in the current working directories, despite the fact that I intend to provide it with a different file.

Edit: Okay, so I'm on an actual keyboard now and I can use words better :D 

I hope the response above explains at least the initial issue; if your default adapter has side effects like file creation, you will get those before you even have the option of configuring them; if they need a network connection, same problem. That'll happen no matter when you do it; you can be trying to bootstrap them in the rails initializer (which I'm now doing, in a careful order) and now no matter what I do I will always have a `flipper.pstore` in the root directory of my application. It's an empty store, but it's misleading, and I cannot avoid it.

I don't think that automatic configuration like this is particularly friendly, tbh; certainly the two days I spent trying to figure out why at random times in my Rails app's lifetime the feature flag provider would change from our actual ActiveRecord one to one configured off of an empty pstore file, sometimes in response to someone else other than me interacting with it was not my favorite experience. Give me _control_ over when things are initialized, please.

**Comment by bkeepers** [comment]
Thanks for the explanation. I really appreciate you taking time to give feedback.

> One reason it’s not so great is that the act of importing the pstore adapter creates a config file in the current working directories, despite the fact that I intend to provide it with a different file.

Let me know if this is not the behavior you are seeing, but the PStore file should not be initialized until the first time the adapter is accessed (by calling `Flipper.enabled?` or something similar). The following does not create an unexpected file for me:

```ruby
# config/initializers/flipper.rb

require "flipper/adapters/pstore"

Flipper.configure do |config|
  config.adapter { Flipper::Adapters::PStore.new("path/to/different/flipper.pstore") }
end
```

<details><summary>See details from `rails console` session to confirm</summary>

```
Loading development environment (Rails 6.1.6.1)
[1] pry(main)> Flipper.configuration.adapter
=> #<Flipper::Adapters::ActiveRecord:0x0000000119f623d0
 @feature_class=Flipper::Adapters::ActiveRecord::Feature (call 'Flipper::Adapters::ActiveRecord::Feature.connection' to establish a connection),
 @gate_class=Flipper::Adapters::ActiveRecord::Gate (call 'Flipper::Adapters::ActiveRecord::Gate.connection' to establish a connection),
 @name=:active_record>
[2] pry(main)> require 'flipper/adapters/pstore'
=> true
[3] pry(main)> Flipper.configuration.adapter
=> #<Flipper::Adapters::PStore:80040 name=:pstore, path="flipper.pstore", store=#<PStore:0x0000000118388d80>>
[4] pry(main)> File.exist?('flipper.pstore')
=> false
[5] pry(main)> Flipper.enabled? :foo
  Flipper feature(foo) enabled? false (0.9ms)  [ thing=nil ]
=> false
[6] pry(main)> File.exist?('flipper.pstore')
=> true
```
</details>

> certainly the two days I spent trying to figure out why…was not my favorite experience.

Really sorry about that. I hate when you're fighting with something for days and it turns out to be something hidden.

We will make it explicit in the docs that loading an adapter makes it the default. At this point we can't remove the automatic configuration because it will cause unexpected failures for other people. We could look into raising a warning or error if multiple adapters are loaded without explicit configuration.

> Give me control over when things are initialized, please.

You definitely have control and are encouraged to explicitly configure flipper (see [example in my first comment here](https://github.com/jnunemaker/flipper/discussions/690#discussioncomment-4461944)) if you're doing anything advanced.

**Comment by offbyone** [comment]
I don't, though! I was dynamically loading the Flipper PStore adapter in a library that was syncing state from activerecord to pstore. Logically, there was no reason to import the pstore adapter anywhere else -- it was used only in that method -- so importing it there, close to the code, made sense... except the part where just importing the code in order to configure it resulted in the reconfiguration of my entire application.

I've found a workaround, so I'm not blocked here, but consider in the future not having system-level side effects when importing code. All hyperbole aside, that's not a developer-friendly practice.
